uid/src/main/java/com/tomtom/kotlin/uid/Uid.kt

```diff
@@ -0,0 +1,225 @@
+        uuid = UUID.randomUUID().toString()
+    }
+
+    /**
+     * Instantiates a Uid with a string. Mainly used when de-serializing existing entities.
+     *
+     * The format of 'uuid' is checked to comply with a standard UUID format, which is:
+     * - Dashes at positions 8, 13, 18, 23 (base 0).
+     * - Characters 0-9 and a-f (lowercase only).
+     *
+     * if this form is used, the creation of the Uid is very fast. If an alternative format
+     * us used, as accepted by [fromString], the call is much more expensive.
+     *
+     * @param uuid Existing string id.
+     * @throws IllegalArgumentException If name does not conform to the string representation
+     * as described in [UUID.toString]. Use [isValid] to make sure the string is valid.
+     */
+    constructor(uuid: String) {
+        /**
+         * This code has been optimized to NOT just call UUID.fromString(uuid) to convert the
+         * UUID-String into a String (and catch an IllegalArgumentException).
+         *
+         * If the UUID does not comply, the expensive call to UUID.fromString is made after all.
+         */
+        val length = uuid.length
+        kotlin.require(
+            UUID_MIN_LENGTH <= length && length <= UUID_MAX_LENGTH
+        ) {
+            "Length of UUID must be [" + UUID_MIN_LENGTH + ", " +
+                UUID_MAX_LENGTH + "], but is " + uuid.length + ", uuid=" + uuid
+        }
+
+        // Check dashes.
+        for (i in UUID_DASH_POS) {
+            if (length < i + 1 || uuid[i] != UUID_DASH) {
+
+                // This will throw an IllegalArgumentException if it went wrong.
+                this.uuid = UUID.fromString(uuid).toString().toLowerCase()
+                return
+            }
+        }
+
+        // Make sure the UUID is lowercase.
+        this.uuid = uuid.toLowerCase()
+
+        // UUID seems to be OK.
+        kotlin.require(onlyContainsValidUUIDCharacters(this.uuid)) {
+            "Incorrect UUID format, uuid=$uuid"
+        }
+    }
+
+    /**
+     * Instantiates an id with a [UUID].
+     *
+     * @param uuid Existing [UUID].
+     */
+    private constructor(uuid: UUID) {
+        this.uuid = uuid.toString()
+    }
+
+    /**
+     * Returns hex string representation of this Uid. Opposite of [fromHexString].
+     * This representation is shorthand, it does not include dashes.
+     *
+     * @return Hex string representation of id, exactly 32 characters long.
+     */
+    fun toHexString(): String {
+        val uuid2 = UUID.fromString(uuid)
+        val msb = java.lang.Long.toHexString(uuid2.mostSignificantBits)
+        val lsb = java.lang.Long.toHexString(uuid2.leastSignificantBits)
+        val sb = StringBuilder()
+        for (i in 1..16 - msb.length) {
+            sb.append('0')
+        }
+        sb.append(msb)
+        for (i in 1..16 - lsb.length) {
+            sb.append('0')
+        }
+        sb.append(lsb)
+        val value = sb.toString()
+        assert(value.length == 32)
```

Comment:
`require`?

Reply:
Not done, must remain assert. The difference between assert and require is that assert expressions are ALWAYS true unless there is a but, but require expressions may be false (which produces an exception).